### ISSUE TITLE
Process vmq_swc_store_meta1 with 0 neighbours crashed with reason: no case clause matching

### ISSUE THREAD

**Issue by dawidslusarczyk1** [vernemq/vernemq#1766]
### Environment

- VerneMQ Version: 1.11.0
- OS: OpenShift 4.6
- VerneMQ configuration (vernemq.conf) set by environment variables:

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: vernemq
  namespace: diiot-dev-poland
  labels:
      app: vernemq
spec:
  serviceName: vernemq
  replicas: 3
  selector:
    matchLabels:
      app: vernemq
  template:
    metadata:
      labels:
        app: vernemq
    spec:
      containers:
      - name: vernemq
        image: image-registry.openshift-image-registry.svc:5000/diiot-dev-poland/vernemq-iiot:1.11.0-ocp
        ports:
        - containerPort: 1883
          name: mqtt
        - containerPort: 8883
          name: mqtts
        - containerPort: 4369
          name: epmd
        - containerPort: 44053
          name: vmq
        - containerPort: 9100
        - containerPort: 9101
        - containerPort: 9102
        - containerPort: 9103
        - containerPort: 9104
        - containerPort: 9105
        - containerPort: 9106
        - containerPort: 9107
        - containerPort: 9108
        - containerPort: 9109
        resources: 
          limits:
            cpu: 500m
            memory: 2Gi
          requests:
            cpu: 10m
            memory: 10Mi
        env:
          - name: MY_POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: DOCKER_VERNEMQ_DISCOVERY_KUBERNETES
            value: '1'
          - name: DOCKER_VERNEMQ_KUBERNETES_SUBDOMAIN
            value: vernemq
          - name: DOCKER_VERNEMQ_KUBERNETES_NAMESPACE
            value: diiot-dev-poland
          - name: DOCKER_VERNEMQ_KUBERNETES_LABEL_SELECTOR
            value: app=vernemq
          - name: DOCKER_VERNEMQ_ERLANG__DISTRIBUTION__PORT_RANGE__MINIMUM
            value: '9100'
          - name: DOCKER_VERNEMQ_ERLANG__DISTRIBUTION__PORT_RANGE__MAXIMUM
            value: '9109'
          - name: DOCKER_VERNEMQ_LISTENER__VMQ__CLUSTERING
            value: '0.0.0.0:44053'
          - name: DOCKER_VERNEMQ_OUTGOING_CLUSTERING_BUFFER_SIZE
            value: '15000000'
          - name: DOCKER_VERNEMQ_METADATA_PLUGIN
            value: vmq_swc
          - name: DOCKER_VERNEMQ_ACCEPT_EULA
            value: 'yes'
          - name: DOCKER_VERNEMQ_ALLOW_ANONYMOUS
            value: 'off'
          - name: DOCKER_VERNEMQ_ALLOW_MULTIPLE_SESSIONS
            value: 'on'
          - name: DOCKER_VERNEMQ_LISTENER.ssl.default
            value: '0.0.0.0:8883'
          - name: DOCKER_VERNEMQ_LISTENER.ssl.cafile
            value: /etc/tls/cacerts.pem
          - name: DOCKER_VERNEMQ_LISTENER.ssl.keyfile
            value: /etc/tls/key.pem
          - name: DOCKER_VERNEMQ_LISTENER.ssl.certfile
            value: /etc/tls/cert.pem
          - name: DOCKER_VERNEMQ_PLUGINS.vmq_passwd
            value: 'off'
          - name: DOCKER_VERNEMQ_PLUGINS.vmq_acl
            value: 'off'
          - name: DOCKER_VERNEMQ_PLUGINS.vmq_diversity
            value: 'on'
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.auth_postgres.enabled
            value: 'on'
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.host
            value: psql-vernemq-authdb.diiot-imel.svc.cluster.local
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.port
            value: '5432'
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.user
            value: vernemq
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.password
            valueFrom:
              secretKeyRef:
                name: vernemq-authdb-credentials
                key: vernemq
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.database
            value: vernemq_db
          - name: DOCKER_VERNEMQ_VMQ_DIVERSITY.postgres.password_hash_method
            value: crypt
          - name: DOCKER_VERNEMQ_LISTENER.ssl.allowed_protocol_versions
            value: '3,4,5,132,131'
          - name: DOCKER_VERNEMQ_PLUGINS.vmq_bridge
            value: 'on'
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.insecure
            value: 'on'
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.max_outgoing_buffered_messages
            value: '1000'
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.restart_timeout
            value: '600'
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.cafile
            value: /etc/tls-clients/mq-broker-diiot-dev-poland.pem
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0
            value: 'mq-broker-diiot-dev-poland.dev.paas.pmiapps.biz:443'
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.client_id
            value: mqttbridge
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.username
            value: mqttbridge
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.password
            valueFrom:
              secretKeyRef:
                name: mqtt-bridge-client-credentials
                key: mqttbridge
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.1
            value: $share/group/PL30/Sensor/Manufacturing/Secondary/* out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.2
            value: $share/group/PL02/Sensor/Manufacturing/Secondary/* out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.3
            value: $share/group/PL02/ProcessOrder/Secondary/* out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.4
            value: $share/group/PL30/ProcessOrder/Secondary/* out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.5
            value: $share/group/empty5 out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.6
            value: $share/group/empty6 out 2
          - name: DOCKER_VERNEMQ_VMQ_BRIDGE.ssl.br0.topic.7
            value: $share/group/empty7 out 2
        volumeMounts:
            - name: vernemq-discovery-node-1
              mountPath: /vernemq/data
            - name: vernemq-discovery-node-2
              mountPath: /vernemq/log
            - name: volume-mq-tls
              mountPath: /etc/tls
            - name: volume-tls-clients
              mountPath: /etc/tls-clients
      volumes:
        - name: vernemq-discovery-node-1
          emptyDir: {}
        - name: vernemq-discovery-node-2
          emptyDir: {}
        - name: volume-mq-tls
          configMap:
            name: mq-tls
            defaultMode: 420
        - name: volume-tls-clients
          configMap:
            name: mq-clietns-cas
            defaultMode: 420
```
- Cluster size/standalone: 3 nodes (Statefull Set pods)

### Expected behavior

- There are no "CRASH" events in the logs

### Actual behavior

- We are facing vernemq cluster node crashing with below messages:

```
03:05:09.374 [error] CRASH REPORT Process vmq_swc_store_batcher_meta1 with 0 neighbours exited with reason: {{{case_clause,{ok,{0,33800}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/app..."},...]},...]},...} in gen_server:call/3 line 223
```
```
03:05:09.374 [error] CRASH REPORT Process vmq_swc_store_meta1 with 0 neighbours crashed with reason: no case clause matching {ok,{0,33800}} in swc_node:event/2 line 186
```
```
03:05:09.375 [error] Supervisor vmq_swc_store_sup_meta1 had child {vmq_swc_store,meta1} started with vmq_swc_store:start_link({swc_config,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local',meta1,vmq_swc_db_meta1,...}) at <0.21155.229> exit with reason no case clause matching {ok,{0,33800}} in swc_node:event/2 line 186 in context child_terminated
```
```
03:05:09.376 [error] Supervisor vmq_swc_store_sup_meta1 had child {vmq_swc_store_batcher,meta1} started with vmq_swc_store_batcher:start_link({swc_config,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local',meta1,vmq_swc_db_meta1,...}) at <0.21162.229> exit with reason {{{case_clause,{ok,{0,33800}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/app..."},...]},...]},...} in context child_terminated
```
This issue is always happening only on one node of the cluster. For now, we aren't really able to reproduce this issue and we are not sure what is causing this.

Below there is also part of the crash.log file but with removed our topics data:
```
2021-03-29 00:14:12 =ERROR REPORT====
** Generic server vmq_swc_store_meta3 terminating 
** Last message in was {MESSAGE PAYLOAD}
** When Server state == {state,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local',undefined,meta3,{swc_config,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local',meta3,vmq_swc_db_meta3,vmq_swc_db_leveldb,vmq_swc_store_meta3,vmq_swc_store_r_o_w_meta3,vmq_swc_store_batcher_meta3,vmq_swc_group_membership_meta3,vmq_swc_edist_srv},{dkm,#Ref<0.1931585748.3445751812.9751>,#Ref<0.1931585748.3445751812.9752>,#Ref<0.1931585748.3445751812.9753>},#{'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => {0,297453748188616851520},'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => {1,56}},{#{'VerneMQ@vernemq-0.vernemq.diiot-dev-poland.svc.cluster.local' => #{'VerneMQ@vernemq-0.vernemq.diiot-dev-poland.svc.cluster.local' => 0,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => 0,'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => 0},'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => #{'VerneMQ@vernemq-0.vernemq.diiot-dev-poland.svc.cluster.local' => 69,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => 69,'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => 69},'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => #{'VerneMQ@vernemq-0.vernemq.diiot-dev-poland.svc.cluster.local' => 7,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => 7,'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => 7}},#{}},['VerneMQ@vernemq-0.vernemq.diiot-dev-poland.svc.cluster.local','VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local'],undefined,true,true,#{},true}
** Reason for termination ==
** {{case_clause,{ok,{0,297453748188616851520}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,266}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,661}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,690}]}]}
** Client vmq_swc_store_batcher_meta3 stacktrace
** [{gen,do_call,4,[{file,"gen.erl"},{line,167}]},{gen_server,call,3,[{file,"gen_server.erl"},{line,219}]},{vmq_swc_store_batcher,loop_accumulating,4,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store_batcher.erl"},{line,50}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]
2021-03-29 00:14:12 =CRASH REPORT====
  crasher:
    initial call: vmq_swc_store:init/1
    pid: <0.20645.227>
    registered_name: vmq_swc_store_meta3
    exception error: {{case_clause,{ok,{0,297453748188616851520}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,266}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,661}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,690}]}]}
    ancestors: [vmq_swc_store_sup_meta3,vmq_swc_sup,<0.215.0>]
    message_queue_len: 0
    messages: []
    links: [<0.241.0>]
    dictionary: [{rand_seed,{#{bits => 58,jump => #Fun<rand.13.8986388>,next => #Fun<rand.10.8986388>,type => exsss,uniform => #Fun<rand.11.8986388>,uniform_n => #Fun<rand.12.8986388>},[139911715735611943|103960466252312268]}}]
    trap_exit: false
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 206432
  neighbours:
2021-03-29 00:14:12 =SUPERVISOR REPORT====
     Supervisor: {local,vmq_swc_store_sup_meta3}
     Context:    child_terminated
     Reason:     {{case_clause,{ok,{0,297453748188616851520}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,266}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,661}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,690}]}]}
     Offender:   [{pid,<0.20645.227>},{id,{vmq_swc_store,meta3}},{mfargs,{vmq_swc_store,start_link,[{swc_config,'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local',meta3,vmq_swc_db_meta3,vmq_swc_db_leveldb,vmq_swc_store_meta3,vmq_swc_store_r_o_w_meta3,vmq_swc_store_batcher_meta3,vmq_swc_group_membership_meta3,vmq_swc_edist_srv}]}},{restart_type,permanent},{shutdown,5000},{child_type,worker}]

2021-03-29 00:14:12 =CRASH REPORT====
  crasher:
    initial call: vmq_swc_store_batcher:init/1
    pid: <0.20661.227>
    registered_name: vmq_swc_store_batcher_meta3
    exception exit: {{{{case_clause,{ok,{0,297453748188616851520}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,266}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,661}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,690}]}]},{gen_server,call,[vmq_swc_store_meta3,{batch,[{{<0.405.0>,#Ref<0.1931585748.3471310849.233225>},{write,[{{{vmq,retain},{[],[MESSAGE PAYLOAD]}">>,#{},undefined}},#{'VerneMQ@vernemq-1.vernemq.diiot-dev-poland.svc.cluster.local' => 69,'VerneMQ@vernemq-2.vernemq.diiot-dev-poland.svc.cluster.local' => 7}}]}}]},infinity]}},[{gen_server,call,3,[{file,"gen_server.erl"},{line,223}]},{vmq_swc_store_batcher,loop_accumulating,4,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store_batcher.erl"},{line,50}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
    ancestors: [vmq_swc_store_sup_meta3,vmq_swc_sup,<0.215.0>]
    message_queue_len: 0
    messages: []
    links: [<0.241.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 319
  neighbours:
2021-03-29 00:14:12 =SUPERVISOR REPORT====
     Supervisor: {local,vmq_swc_store_sup_meta3}
     Context:    child_terminated
     Reason:     {{{case_clause,{ok,{0,297453748188616851520}}},[{swc_node,event,2,[{file,"/opt/vernemq/_build/default/lib/swc/src/swc_node.erl"},{line,186}]},{vmq_swc_store,process_write_op,2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,638}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,'-handle_call/3-fun-1-',2,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,268}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{vmq_swc_store,handle_call,3,[{file,"/opt/vernemq/apps/vmq_swc/src/vmq_swc_store.erl"},{line,266}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,661}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,690}]}]},{gen_server,call,[vmq_swc_store_meta3,{batch,[{{<0.405.0>,#Ref<0.1931585748.3471310849.233225>},{write,[{{{vmq,retain},{[],[MESSAGE PAYLOAD]}
```

**Comment by ioolkos** [comment]
@cloudinittech yeah, it's not reproducible but the likely reason here is that you have node leaves and re-joins with existing state.
It is a bug in SWC syncronisation, and you see it in the `no case clause matching {ok,{0,33800}}`.

0 and 33800 here represent counters for the continuous set and the discontinuous set of events, and for the local node there's an invariant that we should never see `{0, N}` but always `{N, 0}`, that is the events on the local node are always supposed to be continuous. Don't worry, this doesn't have to make immediate sense to you.
I'm working on a fix on some deeper level since a while.

**Comment by cloudinittech** [comment]
@ioolkos Thanks for a quick response 
Is there any workaround? do you recommend not to use the stateful sets?

**Comment by ioolkos** [comment]
You have to use a Stateful set anyway, otherwise Kubernetes doesn't give you persistent node identities.

A possible workaround is implementing/integrating this PR: https://github.com/vernemq/docker-vernemq/pull/264
to tell apart node leaves (as rescaling should use) from node starts (as rescheduling should use)

**Comment by dhruvjain99** [comment]
@ioolkos what's the status on this swc synchronisation bug?

**Comment by ioolkos** [comment]
@dhruvjain99 likely fixed in 1.12.0 with the updates to SWC there.
